controllers/fencer/reconciler.go

```diff
@@ -0,0 +1,184 @@
+	"sigs.k8s.io/controller-runtime/pkg/event"
+
+	"github.com/storageos/api-manager/internal/pkg/cache"
+	"github.com/storageos/api-manager/internal/pkg/storageos"
+)
+
+const (
+	// minPollInterval will override more frequent intervals to protect the
+	// backend api.
+	minPollInterval = 5 * time.Second
+)
+
+var (
+	// ErrNodeNotCached is returned if the node was expected in the cache but
+	// not found.
+	ErrNodeNotCached = errors.New("node not found in cache")
+)
+
+// NodeFencer provides access to nodes and the volumes running on them.
+//go:generate mockgen -destination=mocks/mock_node_fencer.go -package=mocks . NodeFencer
+type NodeFencer interface {
+	ListNodes(ctx context.Context) ([]client.Object, error)
+	GetVolume(ctx context.Context, key client.ObjectKey) (storageos.Object, error)
+}
+
+// Reconciler reconciles StorageOS Node object health with running Pods,
+// deleting them if we know that they are unable to use their storage.
+type Reconciler struct {
+	client.Client
+	scheme          *runtime.Scheme
+	log             logr.Logger
+	api             NodeFencer
+	apiReset        chan<- struct{}
+	pollInterval    time.Duration
+	expiryInterval  time.Duration
+	cleanupInterval time.Duration
+	cache           *cache.Object
+
+	actionv1.Reconciler
+}
+
+// +kubebuilder:rbac:groups="",resources=node,verbs=get;list;watch
+// +kubebuilder:rbac:groups="",resources=pods,verbs=get;list;watch;delete
+// +kubebuilder:rbac:groups="",resources=persistentvolumeclaims,verbs=get;list;watch
+// +kubebuilder:rbac:groups="storage.k8s.io",resources=volumeattachments,verbs=get;list;watch;delete
+
+// NewReconciler returns a new Node label reconciler.
+//
+// The resyncInterval determines how often the periodic resync operation should
+// be run.
+func NewReconciler(api NodeFencer, apiReset chan<- struct{}, k8s client.Client, pollInterval time.Duration, expiryInterval time.Duration) *Reconciler {
+	return &Reconciler{
+		Client:          k8s,
+		log:             ctrl.Log,
+		api:             api,
+		apiReset:        apiReset,
+		pollInterval:    pollInterval,
+		expiryInterval:  expiryInterval,
+		cleanupInterval: 5 * expiryInterval,
+	}
+}
+
+func (r *Reconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager, workers int, retryInterval time.Duration, timeout time.Duration) error {
+	// Create an generic event source. This is used by the Channel type source
+	// to collect the events and process with source event handler.
+	src := make(chan event.GenericEvent)
+
+	// Set scheme for conversions.
+	r.scheme = mgr.GetScheme()
+
+	// Initialize the cache.
+	r.cache = cache.New(r.expiryInterval, r.cleanupInterval)
+
+	// Create an event handler that uses the cache to make reconciliation
+	// decisions.
+	eventHandler := handler.NewEnqueueRequestFromCache(r.cache)
+
+	// Don't allow an overly-agressive interval to break the backend.
+	if r.pollInterval < minPollInterval {
+		r.log.Info("resetting poll interval to minimum", "interval", minPollInterval)
```

Comment:
I guess there is a missing line here, something like `r.pollInterval = minPollInterval`

Reply:
Maybe if this is something which could break the backend some test should be nice to cover it. What do you think?

Reply:
You're right, an assignment is missing.
We don't really know what the control plane is capable of handling. So this is just a vague minimum interval for safety.
I think this conditional can be properly tested by moving this into the constructor `NewReconciler`.